### PR TITLE
Provide clojure-lsp with fallback rootUri and defer starting until we have a Clojure file

### LINKED ISSUE #1601: clojure-lsp does not start in the Getting Started REPL
Clojure-lsp fails to start when firing up a standalone REPL in a folder-less VS Code window.

~I am pretty sure this is a regression cause by~

* #1592

### LINKED ISSUE #1664: clojure-lsp initialized in a non-working state when there is no VS Code folder
VS Code assumes it has a folder open when starting the LSP client. Otherwise it initializes the server with a `null` for `rootUri`. And this leaves clojure-lsp in a non-working state.

This can happen with the standalone REPLs which as a feature can be started standalone from a VS Code folder (see #1601). It can also happen if the user opens Clojure file, or creates a Clojure document. Or, in fact, any of the actions where Calva is configured to activate.

This gets a bit extra bad for the user because the main options for diagnostics we provide **LSP serverInfo**, and **Open LSP log** both crash inside clojure-lsp when the rootUri is not set. And as Calva uses **serverInfo** as part of the activation of clojure-lsp things are a bit in limbo here.

My analysis is that we are starting clojure-lsp prematurely, we need to know the rootUri, or that there is a VS Code folder open, before we start clojure-lsp.

### PR DESCRIPTION
**Pull request by PEZ**
## What has Changed?

VS Code sends `rootUri: null` to the LSP server if there is no folder open in the window. This doesn't work for clojure-lsp. So now, Instead of just blindly starting the server when Calva activates:

* We check if there is a folder open in the window
    * If so we proceed as before
    * If not:
        * We check if we have a Clojure file
           * If we do we start the clojure-lsp server
           * If we don't we don't start it
    * When staring the server
       * We do our best to figure out a fallback rootUri from the file, falling back to a generic fallback if it is untitled.
       * Start the server providing this fallback

The above approach alone would result in no clojure-lsp when starting standalone REPLs. So what we do there is that we: 

1.  wait until we have prepared the ”project” for the standalone REP
2. then we start clojure-lsp.

Fixes #1664 
Fixes #1601

## My Calva PR Checklist

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- ~[ ] Added to or updated docs in this branch, if appropriate~
- [ ] Tests
     - [x] Tested the particular change
     - [x] Figured if the change might have some side effects and tested those as well.
     - [x] Smoke tested the extension as such.
     - [ ] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
     - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
- [x] Created the issue I am fixing/addressing, if it was not present.
- [x] Formatted all JavaScript and TypeScript code that was changed. (use the [prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) or run `npm run prettier-format`)
- [x] Confirmed that there are no linter warnings or errors (use the [eslint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint), run `npm run eslint` before creating your PR, or run `npm run eslint-watch` to eslint as you go).

Ping @pez, @bpringe, @corasaurus-hex, @Cyrik, @ericdallo

### CODE REVIEW

**Comment by ericdallo** [comment]
From my point of view that logic seems to make sense!

**Comment by PEZ** [comment]
@ericdallo thanks for quick help!

The only ”failing” case I've found during my testing is that we don't get any clojure-lsp help on documents which are not on disk, VS Code calls these ”Untitled". But that is regardless if we have a folder/rootUri or not, and actually unrelated to this issue/PR. But anyway. more for the record of things.

**Comment by ericdallo** [comment]
Hum, not sure how hard would be to make that happen, but clojure-lsp should not expect file is on disk indeed

**Comment by PEZ** [comment]
With nREPL, the client sends the file contents. Maybe we can do the same here? Subject for another ticket, though. 😄

**Comment by Cyrik** [comment]
We already send the file contents on open and on change. It's most likely the filename that's making problems

**Comment by PEZ** [comment]
>  It's most likely the filename that's making problems

The file extension?

File: src/lsp/main.ts [hunk 1]
```diff
@@ -350,10 +365,85 @@ async function stopClient() {
...
+        }
+      }
+    }
+  }
+
+  if (!clojureFilePath) {
+    if (!folderType) {
+      folderType = FolderType.FROM_NO_CLOJURE_FILE;
+    }
+    clojureFilePath = path.join(util.calvaTmpDir(), 'clojure-lsp-fallback-root', 'Untitled');
```
Comment (Cyrik) [inline src/lsp/main.ts:421]: Have you tried calling it untitled.clj? That might make it work in lsp

**Comment by Cyrik** [comment]
Yeah, I think lsp ignores anything that doesn't have the right extension

File: src/lsp/main.ts (hunk 1 above)
Comment (PEZ) [reply src/lsp/main.ts:421]: I'm not naming the file here. I'm creating a path on which I can do `dirname`. Maybe needs a clarifying comment...

**Comment by PEZ** [comment]
Untitled files have no extension, by defnition. Seems we can't avoid discussing this off-topic thing here. My fault, for starting it. 😄

**Comment by PEZ** [comment]
I created an issue to discuss the Untitled problem now:

* #1668

**Pull request merged**